src/flashcart/ed64/ed64.c

```diff
@@ -0,0 +1,313 @@
+    // Or somehow warn that certain ROM's will have corruption due to the address space being used for saves.
+
+    if (file_size > (MiB(64) - rom_offset)) {
+        f_close(&fil);
+        return FLASHCART_ERR_ARGS;
+    }
+
+    if (f_lseek(&fil, file_offset) != FR_OK) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    if (f_read(&fil, (void *)(ROM_ADDRESS + rom_offset), file_size, &br) != FR_OK) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+    if (br != file_size) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    if (f_close(&fil) != FR_OK) {
+        return FLASHCART_ERR_LOAD;
+    }
+
+    return FLASHCART_OK;
+}
+
+static flashcart_err_t ed64_load_save (char *save_path) {
+    
+    FIL fil;
+    UINT br;
+
+    if (f_open(&fil, strip_sd_prefix(save_path), FA_READ) != FR_OK) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    size_t save_size = file_get_size(strip_sd_prefix(save_path));
+    uint8_t cartsave_data[save_size];
+
+    if (f_read(&fil, cartsave_data, save_size, &br) != FR_OK) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    if (f_close(&fil) != FR_OK) {
+        return FLASHCART_ERR_LOAD;
+    }
+
+    current_state.is_fram_save_type = false;
+
+    ed64_save_type_t type = ed64_ll_get_save_type();
+    switch (type) {
+        case SAVE_TYPE_EEPROM_4K:
+        case SAVE_TYPE_EEPROM_16K:
+            ed64_ll_set_eeprom(cartsave_data, save_size);
+            break;
+        case SAVE_TYPE_SRAM:
+            ed64_ll_set_sram(cartsave_data, save_size);
+        case SAVE_TYPE_SRAM_128K:
+            ed64_ll_set_sram_128(cartsave_data, KiB(128));
+            break;
+        case SAVE_TYPE_FLASHRAM:
+            ed64_ll_set_fram(cartsave_data, KiB(128));
+            // a cold and warm boot has no way of seeing save types and most types can be determined by size
+            // this tells the cart to use flash instead of sram 128 since they are the same size
+            current_state.is_fram_save_type = true;
+            ed64_state_save(&current_state);
+            break;
+        case SAVE_TYPE_DD64_CART_PORT:
+            break;
+        default:
+            break;
+    }
+
+
+    current_state.last_save_path = save_path;
+    current_state.is_expecting_save_writeback = true;
+    ed64_state_save(&current_state);
+
+    return FLASHCART_OK;
+}
+
+static flashcart_err_t ed64_set_save_type (flashcart_save_type_t save_type) {
+    ed64_save_type_t type;
+
+    switch (save_type) {
+    case FLASHCART_SAVE_TYPE_NONE:
+        type = SAVE_TYPE_NONE;
+        break;
+    case FLASHCART_SAVE_TYPE_EEPROM_4K:
+        type = SAVE_TYPE_EEPROM_4K;
+        break;
+    case FLASHCART_SAVE_TYPE_EEPROM_16K:
+        type = SAVE_TYPE_EEPROM_16K;
+        break;
+    case FLASHCART_SAVE_TYPE_SRAM:
+        type = SAVE_TYPE_SRAM;
+        break;
+    case FLASHCART_SAVE_TYPE_SRAM_BANKED:
```

Comment:
We should add "SRAM banked" support to the `has_feature` function as this is very wrong and won't work with this flashcart. ED64 registers occupy last bank of this save type and that could lead to some undesired behavior. I don't really know right now if we should block loading such save types or just show a warning that this save isn't supported and let user load ROM regardless.

Reply:
I need to check more fully, but the carts would support it, since a) this menu does not write to it. b) if it did, we should store in the pseudo region for flashram.